README.rst
PyJSONProxy - simple proxy and scraper


simple proxy
============

AJAX requests are restricted by the `same origin policy`_. This can be
bypassed by using either `JSONP`_, `CORS`_ or a local proxy. This
implements the third variant. So you can do something like this::

    $ curl http://localhost:5000/github/xi/
    {
      "login": "xi",
      ...
    }

With a configuration like this::

    ENDPOINTS = {
        'github': {
            'host': 'https://api.github.com/users/'
        }
    }


scraping
========

Maybe the more interesting part is that this also contains a simple
scraping mechanism. So if a service does not offer an API but only plain
HTML pages, PyJSONProxy can extract information from there::

    $ curl http://localhost:5000/github/xi/
    {
      "url": "https://github.com/xi/",
      "login": "xi",
      "activity": [
        ...
      ],
      "repos": [{
        ...
      }]
      ...
    }

::

    ENDPOINTS = {
        'github': {
            'host': 'https://github.com/',
            'fields': {
                'login': '.vcard-username',
                'fullname': '.vcard-fullname',
                'email': '.vcard-details .email',
                'join-date': '.vcard-details .join-date@datetime',
                'activity': {
                    'selector': '.contribution-activity-listing ul a'
                },
                'repos': {
                    'selector': '.popular-repos a.mini-repo-list-item',
                    'fields': {
                        'url': '@href',
                        'name': '.repo',
                        'description': '.repo-description'
                    }
                }
            }
        }
    }

Selectors are generally CSS-selectors with the additional option to
select an attribute by appending an ``@`` and the attribute name. If no
attribute is selected, the text content of the element will be used.


CORS header
===========

By setting ``ALLOW_CORS`` to ``True``, an
``Access-Control-Allow-Origin``-header with value ``*`` will be set for
all responses.


Documentation
=============

Some simple documentation is automatically generated and available under
``/`` (for all endpoints) or ``/{endpoint}/`` (for an individual
endpoint). To provide some input for this documentation, you can add a
description to both endpoints and fields::

    ENDPOINTS = {
        'github': {
            'host': 'https://github.com/',
            'doc': 'Access data about GitHub users',
            'fields': {
              'login': '.vcard-username',
              'fullname': '.vcard-fullname',
              'email': '.vcard-details .email'
              'join-date': '.vcard-details .join-date@datetime'
            },
            'fields_doc': {
              'login': 'github username',
              'fullname': 'the user\'s full name',
              'join-date': 'date when the user joined github in ISO-xx format'
            }
        }
    }


Note on security and performance
================================

Security and performance were not a priority in this project. So it
might be bad.


.. _same origin policy: https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy
.. _JSONP: https://en.wikipedia.org/wiki/JSONP
.. _CORS: https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS


CHANGES.rst
Changes
=======

2.1.0 (2019-05-05)
------------------

- fix regression: non-scraping requests
- use aiohttp directly instead of deprecated fakes
- define __version__


2.0.0 (2017-06-03)
------------------

- drop support for python3.4
- handle HEAD and OPTIONS requests internally instead of forwarding it
- allow to mark fields as optional by appending a '?' to the selector
- allow simple formatting of host names
- include template for documentation


1.0.2 (2015-12-22)
------------------

- fix distribution


1.0.1 (2015-12-06)
------------------

- some small fixes
- a larger part has been moved to a separate project:
  `Fakes <https://github.com/xi/fakes>`_


1.0.0 (2015-12-06)
------------------

Large parts have been reimplemented with asyncio.  This results in much
improved performance, but support for python2 has been dropped.  This release
should be compatible with the last one.


0.1.0 (2015-12-04)
------------------

- breaking change: more flexible scrape configuration


0.0.1 (2015-06-04)
------------------

- fix distribution


0.0.0
-----

initial release


jsonproxy/api.py
from __future__ import absolute_import

try:
	from urllib.request import urlopen as _urlopen
	from urllib.error import HTTPError
except ImportError:
	from urllib2 import urlopen as _urlopen
	from urllib2 import HTTPError

from flask import Blueprint
from flask import request
from flask import current_app
from flask import abort
from flask import jsonify
from flask import make_response
from flask import render_template
import cachetools

from .lib import _doc
from .lib import ENDPOINTS
from .lib import scrape

api = Blueprint('api', __name__, static_folder='static')


def get_config(endpoint):
	try:
		return current_app.config[ENDPOINTS][endpoint]
	except KeyError:
		abort(404)


@cachetools.ttl_cache()
def urlopen(url):
	try:
		current_app.logger.info('fetching %s' % url)
		original = _urlopen(url)

		body = original.read()
		code = original.getcode()
		headers = original.headers.items()

		return body, code, headers
	except HTTPError as error:
		abort(error.code)


@api.route('/<endpoint>/<path:path>', methods=['GET'])
def main(endpoint, path):
	config = get_config(endpoint)
	url = request.url.replace(request.host_url + endpoint + '/', config['host'])

	body, code, headers = urlopen(url)

	if 'fields' in config:
		if code == 200:
			response = jsonify(scrape(url, body, config))
		else:
			abort(code)
	else:
		response = make_response(body, code)

	if current_app.config.get('ALLOW_CORS', False):
		response.headers['Access-Control-Allow-Origin'] = '*'

	return response


@api.route('/', methods=['GET'])
def index():
	config = current_app.config[ENDPOINTS]
	data = [_doc(config[endpoint], endpoint) for endpoint in config]
	return render_template('index.html', endpoints=data)


@api.route('/<endpoint>/', methods=['GET'])
def doc(endpoint):
	config = get_config(endpoint)
	return render_template('index.html', endpoints=[_doc(config, endpoint)])


jsonproxy/lib.py
from bs4 import BeautifulSoup

try:
	from functools import lru_cache
except ImportError:
	from cachetools import lru_cache

ENDPOINTS = 'ENDPOINTS'


def iter_attribute(html, selector):
	optional = selector.endswith('?')
	if optional:
		selector = selector[:-1]

	s = selector.rsplit('@', 1)[0]
	if s:
		elements = html.select(s)
	else:
		elements = [html]

	if '@' in selector:
		attr = selector.rsplit('@', 1)[1]
		if optional:
			return (element.get(attr) for element in elements)
		else:
			return (element[attr] for element in elements)
	else:
		return (element.text.strip() for element in elements)


def get_attribute(html, selector):
	iterator = iter_attribute(html, selector)
	try:
		return next(iterator)
	except StopIteration:
		if not selector.endswith('?'):
			raise


def get_fields(html, config):
	data = {}
	for key, value in config['fields'].items():
		if isinstance(value, str):
			data[key] = get_attribute(html, value)
		elif 'fields' in value:
			elements = html.select(value['selector'])
			data[key] = [get_fields(e, value) for e in elements]
		else:
			data[key] = list(iter_attribute(html, value['selector']))
	return data


@lru_cache()
def parse_html(body):
	return BeautifulSoup(body)


def scrape(url, body, config):
	html = parse_html(body)
	data = get_fields(html, config)
	data['url'] = url
	if 'post' in config:
		data = config['post'](data)
	return data


def _fields_doc(config):
	if isinstance(config, dict):
		fields = config.get('fields', {})
		doc = config.get('fields_doc', {})
		for key in fields:
			yield key, doc.get(key, ''), list(_fields_doc(fields[key]))


def _doc(config, endpoint):
	data = {
		'title': endpoint,
		'doc': config.get('doc', ''),
		'fields': list(_fields_doc(config)),
	}

	data['fields'].append(('url', 'url of the scraped page', []))

	return data


def check_fields_config(fields, endpoint, field=''):
	for key, value in fields.items():
		full_key = field + '.' + key if field else key
		if isinstance(value, dict):
			if 'selector' not in value:
				yield ('No selector configured for field %s in endpoint %s.' % (
					full_key, endpoint
				))
			if 'fields' in value:
				for error in check_fields_config(value['fields'], endpoint, full_key):
					yield error


def check_config(config):
	errors = []

	if ENDPOINTS not in config or len(config[ENDPOINTS]) == 0:
		errors.append('No endpoints configured.')
	else:
		for key, data in config[ENDPOINTS].items():
			if 'fields' in data:
				if len(data['fields']) == 0:
					errors.append('No fields configured for endpoint %s.' % key)
				else:
					errors += list(check_fields_config(data['fields'], key))

	return errors


jsonproxy/web.py
"""Flask inspired wrapper around aiohttp."""

from functools import lru_cache
from pkg_resources import resource_filename
import asyncio
import logging
import os

from aiohttp import web
import jinja2


@lru_cache()
def get_template(name):
	path = resource_filename(__name__, os.path.join('templates', name))
	with open(path) as fh:
		return jinja2.Template(fh.read())


def render_template(name, **kwargs):
	template = get_template(name)
	text = template.render(**kwargs)
	return web.Response(body=text.encode('utf8'))


def jsonify(data, **kwargs):
	return web.json_response(data, **kwargs)


def abort(code):
	if code == 404:
		raise web.HTTPNotFound
	elif code >= 500:
		raise web.HTTPInternalServerError
	else:
		raise web.HTTPBadRequest


def make_response(data, **kwargs):
	if isinstance(data, web.StreamResponse):
		return data
	elif isinstance(data, str):
		return web.Response(body=data.encode('utf8'), **kwargs)
	elif isinstance(data, bytes):
		return web.Response(body=data, **kwargs)
	else:
		raise TypeError('cannot make response from {}'.format(data))


class Application:
	def __init__(self, name):
		self.name = name
		self.loop = asyncio.get_event_loop()
		self.app = web.Application(loop=self.loop)
		self.config = {}
		self.debug = False

		self.logger = logging.getLogger(self.name)
		self.logger.setLevel(logging.INFO)

		consoleHandler = logging.StreamHandler()
		formatter = logging.Formatter('%(asctime)s - %(message)s')
		consoleHandler.setFormatter(formatter)
		self.logger.addHandler(consoleHandler)

	def config_from_file(self, path):
		with open(path) as fh:
			exec(compile(fh.read(), path, 'exec'), self.config)

	def add_route(self, path, fn, methods=('GET',)):
		@asyncio.coroutine
		def wrapped(*args, **kwargs):
			if asyncio.iscoroutinefunction(fn):
				data = yield from fn(*args, **kwargs)
			else:
				data = fn(*args, **kwargs)
			return make_response(data)

		for method in methods:
			self.app.router.add_route(method, path, wrapped)

	def route(self, path, methods=('GET',)):
		def decorator(fn):
			self.add_route(path, fn, methods=methods)
			return fn
		return decorator

	def run(self, host='localhost', port=5000):
		if self.debug:
			self.logger.setLevel(logging.DEBUG)
		server = self.loop.create_server(self.app.make_handler(), host, port)
		self.loop.run_until_complete(server)
		self.logger.info("Server started at http://{}:{}".format(host, port))
		try:
			self.loop.run_forever()
		except KeyboardInterrupt:
			pass


jsonproxy/__init__.py
from functools import lru_cache
from pkg_resources import resource_filename
import argparse
import asyncio
import os
import sys

from aiohttp import web
import aiohttp
import jinja2

from .lib import check_config
from .lib import _doc
from .lib import ENDPOINTS
from .lib import scrape

__version__ = '2.1.0'

CONFIG = {}


def get_config(endpoint):
	try:
		return CONFIG[ENDPOINTS][endpoint]
	except KeyError:
		raise aiohttp.web_exceptions.HTTPNotFound


@lru_cache()
def get_template(name):
	local_path = os.path.join('templates', name)
	path = resource_filename(__name__, local_path)
	with open(path) as fh:
		s = fh.read()
		return jinja2.Template(s)


def render_template(name, **kwargs):
	"""Shortcut for rendering a jinja template to a response."""
	template = get_template(name)
	text = template.render(**kwargs)
	return web.Response(
		body=text.encode('utf8'),
		content_type='text/html')


def async_cache(maxsize=128):
	cache = {}

	def decorator(fn):
		async def wrapper(*args):
			key = ':'.join(args)
			if key not in cache:
				if len(cache) >= maxsize:
					del cache[cache.keys().next()]
				cache[key] = await fn(*args)
			return cache[key]
		return wrapper
	return decorator


@async_cache()
async def _request(method, url):
	async with aiohttp.request(method, url) as response:
		if response.status == 404:
			raise aiohttp.web_exceptions.HTTPNotFound
		response.raise_for_status()
		# get response before closing the connection
		body = await response.read()
		return body, response


async def handle(request):
	endpoint = request.match_info['endpoint']

	config = get_config(endpoint)
	if '{' in config['host']:
		parts = request.match_info['path'].strip('/').split('/')
		url = config['host'].format(*parts)
	else:
		url = config['host'] + request.match_info['path']
	if request.query_string:
		url += '?' + request.query_string

	body, original = await _request(request.method, url)

	if 'fields' in config:
		data = scrape(url, body, config)
		response = web.json_response(data, status=original.status)
	else:
		response = web.Response(
			body=body, status=original.status, content_type=original.content_type
		)

	if CONFIG.get('ALLOW_CORS', False):
		response.headers['Access-Control-Allow-Origin'] = '*'

	return response


def index(request):
	config = CONFIG[ENDPOINTS]
	data = [_doc(config[endpoint], endpoint) for endpoint in config]
	return render_template('index.html', endpoints=data)


def doc(request):
	endpoint = request.match_info['endpoint']
	config = get_config(endpoint)
	data = [_doc(config, endpoint)]
	return render_template('index.html', endpoints=data)


def parse_args():
	parser = argparse.ArgumentParser(description='simple proxy and scraper')
	parser.add_argument('config')
	parser.add_argument('--version', action='version', version=__version__)
	parser.add_argument('-d', '--debug', action='store_true')
	parser.add_argument('-p', '--port', type=int, default=5000)
	parser.add_argument('-H', '--host', default='localhost')
	return parser.parse_args()


def main():
	args = parse_args()

	config_path = os.path.abspath(args.config)
	with open(config_path) as fh:
		exec(compile(fh.read(), config_path, 'exec'), CONFIG)  # nosec

	errors = check_config(CONFIG)
	if errors:
		for error in errors:
			print(error)
		sys.exit(1)

	loop = asyncio.get_event_loop()
	app = web.Application(loop=loop)

	app.router.add_route('GET', '/', index)
	app.router.add_route('GET', '/{endpoint}/', doc)
	app.router.add_route('GET', '/{endpoint}/{path:.+}', handle)

	web.run_app(app, host=args.host, port=args.port)


if __name__ == '__main__':
	main()
